sdk/testing/src/test/java/io/opentelemetry/sdk/testing/context/StrictContextStorageTest.java

```diff
@@ -0,0 +1,181 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.opentelemetry.sdk.testing.context;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.TraceFlags;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.ContextKey;
+import io.opentelemetry.context.ContextStorage;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.sdk.trace.IdGenerator;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings("MustBeClosedChecker")
+class StrictContextStorageTest {
+
+  private static final ContextKey<String> ANIMAL = ContextKey.named("animal");
+  private static final IdGenerator IDS_GENERATOR = IdGenerator.random();
+  private static final String TRACE_ID = IDS_GENERATOR.generateTraceId();
+  private static final String SPAN_ID = IDS_GENERATOR.generateSpanId();
+
+  private static final Span SPAN =
+      Span.wrap(
+          SpanContext.create(TRACE_ID, SPAN_ID, TraceFlags.getSampled(), TraceState.getDefault()));
+
+  private static ContextStorage previousStorage;
+  private static StrictContextStorage strictStorage;
+
+  @BeforeAll
+  static void setUp() {
+    previousStorage = SettableContextStorageProvider.getContextStorage();
+    strictStorage = StrictContextStorage.create(previousStorage);
+    SettableContextStorageProvider.setContextStorage(strictStorage);
+  }
+
+  // In this test we intentionally leak context so need to restore it ourselves, bypassing the
+  // strict storage.
+  @AfterEach
+  void resetContext() {
+    previousStorage.attach(Context.root());
+  }
+
+  @AfterAll
+  static void tearDown() {
+    SettableContextStorageProvider.setContextStorage(previousStorage);
+  }
+
+  // TODO(anuraaga): These rules conflict with error prone so one or the other needs to be
+  // disabled.
+  @SuppressWarnings({"checkstyle:EmptyBlock", "checkstyle:WhitespaceAround"})
+  @Test
+  void decorator_close_afterCorrectUsage() {
+    try (Scope ws = Context.current().with(ANIMAL, "cat").makeCurrent()) {
+      try (Scope ws2 = Context.current().with(ANIMAL, "dog").makeCurrent()) {}
+    }
+
+    strictStorage.ensureAllClosed(); // doesn't error
+  }
+
+  @Test
+  public void doesntDecorateNoop() {
```

Comment:
Wouldn't we want to decorate the no-op scope, to make sure that even those are being properly closed? I can imagine use-cases where someone would accidentally end up creating a no-op scope in a test, but in the real world it would end up being a real scope, and then we wouldn't flag it as a problem.

Reply:
Yeah that makes sense, removed the noop handling